packages/react-native/ReactCommon/react/renderer/textlayoutmanager/platform/ios/react/renderer/textlayoutmanager/RCTFontUtils.mm

```diff
@@ -14,6 +14,61 @@
 #import <map>
 #import <mutex>
 
+static UIFontWeight weightOfFont(UIFont *font)
+{
+  static NSArray<NSString *> *weightSuffixes;
+  static NSArray<NSNumber *> *fontWeights;
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    // We use two arrays instead of one map because
+    // the order is important for suffix matching.
+    weightSuffixes = @[
+      @"normal",
+      @"ultralight",
+      @"thin",
+      @"light",
+      @"regular",
+      @"medium",
+      @"semibold",
+      @"demibold",
+      @"extrabold",
+      @"ultrabold",
+      @"bold",
+      @"heavy",
+      @"black"
+    ];
+    fontWeights = @[
+      @(UIFontWeightRegular),
+      @(UIFontWeightUltraLight),
+      @(UIFontWeightThin),
+      @(UIFontWeightLight),
+      @(UIFontWeightRegular),
+      @(UIFontWeightMedium),
+      @(UIFontWeightSemibold),
+      @(UIFontWeightSemibold),
+      @(UIFontWeightHeavy),
+      @(UIFontWeightHeavy),
+      @(UIFontWeightBold),
+      @(UIFontWeightHeavy),
+      @(UIFontWeightBlack)
+    ];
+  });
+
+  NSString *fontName = font.fontName;
+  NSInteger i = 0;
+  for (NSString *suffix in weightSuffixes) {
+    // CFStringFind is much faster than any variant of rangeOfString: because it does not use a locale.
+    auto options = kCFCompareCaseInsensitive | kCFCompareAnchored | kCFCompareBackwards;
+    if (CFStringFind((CFStringRef)fontName, (CFStringRef)suffix, options).location != kCFNotFound) {
+      return fontWeights[i].doubleValue;
+    }
+    i++;
+  }
```

Comment:
we probably don't need a loop if we can use a dictionary approach. We might have to extract the suffix from the font, but it would be better and more readable.